src/data_source/fetching.rs

```diff
@@ -900,6 +908,12 @@ where
             vid_common_fetcher = vid_common_fetcher.with_retry_delay(delay);
         }
 
+        if let Some(limit) = builder.rate_limit {
+            payload_fetcher = payload_fetcher.with_rate_limit(limit);
+            leaf_fetcher = leaf_fetcher.with_rate_limit(limit);
+            vid_common_fetcher = vid_common_fetcher.with_rate_limit(limit);
```

Comment:
I'm thinking instead of `limit` each fetcher should take an `Arc<Semaphore>`, so we can share the same semaphore between all of them. That way instead of a rate limit *per resource type* (which is kind of arbitrary) we just have an overall limit